Fix dirty tracking after rollback.

kamipo opened the pull request:
Currently the rollback only restores primary key value, `new_record?`,
`destroyed?`, and `frozen?`. Since the `save` clears current dirty
attribute states, retrying save after rollback will causes no change
saved if partial writes is enabled (by default).

This makes `remember_transaction_record_state` remembers original values
then restores dirty attribute states after rollback.

Fixes #15018.
Fixes #30167.
Fixes #33868.
Fixes #33443.
Closes #33444.
Closes #34504.

kamipo committed:
Fix dirty tracking after rollback.

Currently the rollback only restores primary key value, `new_record?`,
`destroyed?`, and `frozen?`. Since the `save` clears current dirty
attribute states, retrying save after rollback will causes no change
saved if partial writes is enabled (by default).

This makes `remember_transaction_record_state` remembers original values
then restores dirty attribute states after rollback.

Fixes #15018.
Fixes #30167.
Fixes #33868.
Fixes #33443.
Closes #33444.
Closes #34504.

kaspth commented:
Nice! Is there any performance impact of running `sync_with_transaction_state` more often now?

kamipo commented:
https://gist.github.com/kamipo/cb6138e01f2f706c4f2e87a1282c8428

Before (d4e2824d8d9aea6d95e78275c0107bbbc48300d9 before #35933):

```
Warming up --------------------------------------
             user.id    82.135k i/100ms
          user['id']    63.284k i/100ms
           user.name   154.252k i/100ms
        user['name']    92.613k i/100ms
       user.changed?    10.552k i/100ms
 user.saved_changes?    25.754k i/100ms
Calculating -------------------------------------
             user.id      1.101M (±10.9%) i/s -      5.421M in   5.045868s
          user['id']    794.397k (± 1.0%) i/s -      3.987M in   5.019332s
           user.name      2.637M (± 2.5%) i/s -     13.266M in   5.033214s
        user['name']      1.276M (± 1.5%) i/s -      6.390M in   5.009238s
       user.changed?    111.988k (± 9.7%) i/s -    559.256k in   5.052791s
 user.saved_changes?    278.738k (±12.3%) i/s -      1.365M in   5.018580s
```

After (this PR 63ff495bdf90e0ab20114a49db5cffe3cb9ef2fd):

```
Warming up --------------------------------------
             user.id   118.536k i/100ms
          user['id']    94.816k i/100ms
           user.name   151.977k i/100ms
        user['name']    92.442k i/100ms
       user.changed?    18.335k i/100ms
 user.saved_changes?    24.328k i/100ms
Calculating -------------------------------------
             user.id      2.108M (± 1.7%) i/s -     10.550M in   5.007100s
          user['id']      1.294M (± 1.1%) i/s -      6.542M in   5.057655s
           user.name      2.544M (± 1.0%) i/s -     12.766M in   5.019074s
        user['name']      1.290M (± 1.0%) i/s -      6.471M in   5.018621s
       user.changed?    191.417k (±11.6%) i/s -    953.420k in   5.097952s
 user.saved_changes?    263.707k (±12.1%) i/s -      1.289M in   5.012170s
```

`user.id` is 2x faster by 8180c3961063c5786305a94a7d204ab4bc72c381.
Perhaps `user['name']` is a bit faster and `user.name` is a bit slower, but not significant.

kaspth commented:
Great, thanks for checking! 🤘

shioyama commented:
Since this overlaps with parts of the codebase I'm looking at right now, I decided to delve into the changes here and especially confirm performance numbers because they didn't really make sense to me.

And indeed, I'm getting different ones. I compared directly before and directly after this commit, and here's what I get:

Before (315dcb6751c8e10f40189a9838f418e06df45560)

```
Warming up --------------------------------------
             user.id   175.328k i/100ms
          user['id']   125.725k i/100ms
           user.name   225.567k i/100ms
        user['name']   133.138k i/100ms
       user.changed?    25.216k i/100ms
 user.saved_changes?    36.525k i/100ms
Calculating -------------------------------------
             user.id      2.497M (± 1.9%) i/s -     12.624M in   5.057201s
          user['id']      1.610M (± 2.1%) i/s -      8.172M in   5.077406s
           user.name      3.466M (± 2.0%) i/s -     17.369M in   5.014104s
        user['name']      1.685M (± 2.1%) i/s -      8.521M in   5.059298s
       user.changed?    266.509k (± 9.9%) i/s -      1.311M in   5.007683s
 user.saved_changes?    362.952k (±16.0%) i/s -      1.753M in   5.049761s
```

After (ef21edbc45275f3be434732d14a284980a1a291b)

```
Warming up --------------------------------------
             user.id   181.817k i/100ms
          user['id']   130.409k i/100ms
           user.name   208.120k i/100ms
        user['name']   132.942k i/100ms
       user.changed?    24.702k i/100ms
 user.saved_changes?    34.175k i/100ms
Calculating -------------------------------------
             user.id      2.641M (± 2.1%) i/s -     13.273M in   5.027979s
          user['id']      1.697M (± 2.1%) i/s -      8.607M in   5.072999s
           user.name      3.128M (± 2.0%) i/s -     15.817M in   5.058316s
        user['name']      1.685M (± 1.9%) i/s -      8.508M in   5.052189s
       user.changed?    260.651k (±10.0%) i/s -      1.285M in   5.017188s
 user.saved_changes?    354.439k (±13.5%) i/s -      1.777M in   5.166725s
```

I *do* see an improvement of `user.id`, but not by a factor of 2x, more like around 5% or so.

But more importantly, I see a more significant drop in performance of the getter method `user.name`, which is degraded by more than 10%. And this makes sense because `sync_with_transaction_state` is being called now on every fetch, whereas previously it was left out of the getter method entirely (so never needed to even call out to the method). I also see smaller drops in performance to the other dirty methods, I assume for the same reason.

I believe the reason @kamipo you're seeing the 2x boost is because your commit span includes (your commit) f7e126beae6249ad75d94bc585f53234b8ae5016. When I benchmark that change, I indeed see that much greater boost in `user.id`.

Your span from d4e2824...63ff495 includes f7e126beae6249ad75d94bc585f53234b8ae5016 in addition to the dirty change here, so although there *has* been a huge boost in performance of `user.id`, I don't believe it is attributable to *this* change.

It would be great if someone else could double-check the results, because with all these commits flying around, it's easy to make a mistake. (I used `user.private_methods.include?(:thaw)` as a sanity check to confirm which side of the commit I was on.)

If my suspicion is correct, then we are paying a 10% hit on every getter in order to restore dirty tracking state on rollback, which seems too expensive, but at least we should be aware of that cost.

shioyama commented:
Sorry, I see now that @kamipo's comment mentions #35933. But the point remains then that `user.id` is not improved significantly by this change, whereas `user.name` performance drops, I think significantly but that needs to be confirmed.

shioyama commented:
The commit that gives (most of?) the 2x boost is I believe 8180c3961063c5786305a94a7d204ab4bc72c381.

I might be totally not understanding something here, if so I apologize.

shioyama commented:
Maybe I'm not understanding the history of this change, if so I apologize.
